designs/2022-languages/README.md

```diff
@@ -0,0 +1,845 @@
+
+We did also previously accept [#56](https://github.com/eslint/rfcs/pull/56) as an extension to `parseForESLint()` to support more languages and [eslint/eslint#14745](https://github.com/eslint/eslint/issues/14745) has been open for a couple of years to figure out a way to lint both a file and its virtual parts.
+
+## Detailed Design
+
+This proposal consists of the following changes:
+
+1. Allows plugins to publish languages in addition to rules, processors, etc.
+1. Allows end users to select which language to use in their config files.
+1. Defines an API for defining languages.
+1. Changes the ESLint core to consume language definitions instead of assuming JavaScript is always used.
+
+**Prerequisites:** This proposal requires the flat config system and cannot be used with the eslintrc config system.
+
+**Note:** This design is intended to work both in the current version of ESLint and also in the "rewrite" version, which I'll refer to as ESLint X to distinguish between the two. Not all features of this proposal will be available in both versions. 
+
+### Languages in plugins and configs
+
+Languages can be defined in plugins by using the `languages` key and giving the language a name. This allows multiple languages to be contained in a single plugin if desired. For example:
+
+```js
+// inside a plugin called eslint-plugin-lang
+export default {
+
+    languages: {
+        lang1: {
+            // language object
+        },
+        lang2: {
+            // language object
+        }
+    },
+    processors: {
+        // ..
+    },
+    rules: {
+        // ...
+    }
+};
+```
+
+Here, `lang1` and `lang2` are languages defined by the plugin. End users can specify which language to use via the `language` key in a config object, such as:
+
+
+```js
+// inside eslint.config.js
+import lang from "eslint-plugin-lang";
+
+export default [
+    {
+        plugins: {
+            lang
+        },
+        files: ["**/*.lang"],
+        language: "lang/lang1",
+        languageOptions: {
+            // options
+        }
+    }
+];
+```
+
+Here, `"lang/lang1"` refers to the `lang1` language as exported from `eslint-plugin-lang`. Anything inside of `languageOptions` is then handled by `lang1` rather than by ESLint itself. 
+
+### Language Definition Object
+
+Each language definition object must implement the following interface:
+
+```ts
+interface ESLintLanguage {
+
+    options: {
+
+        /**
+         * Indicates how ESLint should treat the file.
+         */
+        fileType: "text" | "binary";
+
+        /**
+         * First line number returned from the parser (text mode only).
+         */
+        lineStart?: 0 | 1;
+
+        /**
+         * First column number returned from the parser (text mode only).
+         */
+        columnStart?: 0 | 1;
+        
+        /**
+         * Schema for language options.
+         */
+        languageOptionsSchema: ObjectSchemaDefinitions;
```

Comment:
@mdjermanovic After looking into the implementation more, this won't work. ConfigArray would need to merge the `language` key, get the result, and then use that before `languageOptions` key was merged in order to use `merge()`. There's just no way to do that. 

As such, I think the best way forward is to have a `validateOptions()` method that is called not during ConfigArray merging but rather just before methods are called on the language.

Reply:
Ah, yes. What shall we do with the `languageOptions` [schema](https://github.com/eslint/eslint/blob/892e6e58c5a07a549d3104de3b6b5879797dc97f/lib/config/flat-config-schema.js#L431-L439) then? Currently, It's hard-coded for the JS language.

The [default](https://github.com/eslint/eslint/blob/892e6e58c5a07a549d3104de3b6b5879797dc97f/lib/config/default-config.js#L40-L45) `languageOptions` are also hard-coded, so those keys will appear in merged configs for other languages.

There's also some js language-specific logic about `languageOptions` [in finalizeConfig](https://github.com/eslint/eslint/blob/892e6e58c5a07a549d3104de3b6b5879797dc97f/lib/config/flat-config-array.js#L191-L205).